### PR TITLE
[wasm] Wasm.Build.Tests - some refactoring, and rationalizing

### PR DESCRIPTION
**Pull request by radical**
This adds new `*ProjectProvider` classes that will have methods for project-type specific things like expected list of `dotnet.*` files, methods to create/run projects. This is the main new change here. A streamlined way of knowing which files we are looking for in tests.

- This removes a check for `isNet7AndBelow` which was unused. But in a follow up PR, blazor `net7.0` project will be added
- `BuildTestBase.cs` has lot of project-type specific code, but almost none of that has been moved to the provider classes, in an effort to keep this PR small.
- Other changes are for making long parameter lists into `Options` classes.
- There is some code for `multithreaded` projects, but it isn't used as yet, but will get utilized in a follow up PR.
- More cleanup, and moving code around will happen in the follow up PR

### CODE REVIEW

**Comment by radical** [comment]
I recommend looking at this commit wise, which should make reviewing a little easier.

**Review by ilonatommy — APPROVED** [review]
Looks very good

File: src/mono/wasm/Wasm.Build.Tests/BlazorWasmProjectProvider.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+#nullable enable
+
+namespace Wasm.Build.Tests;
+
+public class BlazorWasmProjectProvider(string projectDir, ITestOutputHelper _testOutput)
+                : ProjectProviderBase(projectDir, _testOutput)
+{
+    public void AssertBlazorBootJson(string config, bool isPublish, bool isNet7AndBelow, string targetFramework, string? binFrameworkDir)
+    {
```
Comment (maraf) [inline src/mono/wasm/Wasm.Build.Tests/BlazorWasmProjectProvider.cs:26]: Blazor boot json schema is now shared with WasmAppBuilder. We should do the same check for every app bundle

File: src/mono/wasm/Wasm.Build.Tests/TestMainJsProjectProvider.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+using Xunit.Abstractions;
+
+namespace Wasm.Build.Tests;
+
+public class TestMainJsProjectProvider(string projectDir, ITestOutputHelper testOutput)
+                : ProjectProviderBase(projectDir, testOutput)
+{
+    // no fingerprinting
+    protected override IReadOnlyDictionary<string, bool> GetDotnetFilesKnownSet(RuntimeVariant runtimeType)
+        => new SortedDictionary<string, bool>()
```
Comment (maraf) [inline src/mono/wasm/Wasm.Build.Tests/TestMainJsProjectProvider.cs:17]: Could we merge `GetDotnetFilesKnownSet` and `GetDotNetFilesExpectedSet` and provide a collection of records (with properties)?

Comment (maraf) [reply src/mono/wasm/Wasm.Build.Tests/TestMainJsProjectProvider.cs:17]: And maybe a second thought. If I image that I add a new runtime file, I would need to find all places and add the file there. What I like about similar SDK tests, is that in such a case, I can just run all tests in mode "generate baselines" and than check the generated changes in git diff.

**Review by maraf — COMMENTED** [review]
It is a good improvement! I have few comments/ideas.

File: src/mono/wasm/Wasm.Build.Tests/TestMainJsProjectProvider.cs (hunk 2 above)
Comment (radical) [reply src/mono/wasm/Wasm.Build.Tests/TestMainJsProjectProvider.cs:17]: The baselines stuff has been a major pain in sdk tests, and it's already well covered there. So I don't think we need to duplicate that here. One of the reasons for moving to this new way of testing for dotnet files was so the logic is all shared, and the difference is just the list of files. And the reason for difference list is that we still have different kinds of builds - blazor, templates, and test-main-js based tests, each having differences in their bundles. And these will gradually converge.

What I can do is try to ensure that a new runtime file would have to be added in only these lists. Does that make sense?

File: src/mono/wasm/Wasm.Build.Tests/BlazorWasmProjectProvider.cs (hunk 1 above)
Comment (radical) [reply src/mono/wasm/Wasm.Build.Tests/BlazorWasmProjectProvider.cs:26]: And the same thing is needed for checking icu files.

File: src/mono/wasm/Wasm.Build.Tests/TestMainJsProjectProvider.cs (hunk 2 above)
Comment (maraf) [reply src/mono/wasm/Wasm.Build.Tests/TestMainJsProjectProvider.cs:17]: I don't want to duplicate baseline tests from SDK, I'm just looking for a good way to discover all places for scenario when a new runtime file is added.

Comment (maraf) [reply src/mono/wasm/Wasm.Build.Tests/TestMainJsProjectProvider.cs:17]: > Could we merge GetDotnetFilesKnownSet and GetDotNetFilesExpectedSet and provide a collection of records (with properties)?

Do we need to method with two lists of files? The lists look identical. What am I missing?

Comment (radical) [reply src/mono/wasm/Wasm.Build.Tests/TestMainJsProjectProvider.cs:17]: One is the list of all the known `dotnet.*` files to expect for a given project type, and whether to expect them to be fingerprinted. The other is for the actual list expected for individual tests. I can look at possibly improving this in the follow up PR.

**Review by maraf — APPROVED** [review]

**Pull request merged**